### PR TITLE
Change target branch for pull request

### LINKED ISSUE #6377: [Feature request] Pull requests: Allow to change target branch
When creating a new Pull request you decide which branch will be the target of the pull request. 

After creating it, it is not possible to change the target. 
In case it has to be change, you have to close the pull request including all notes etc. 

It would be a nice feature, if the user is able to change the target branch.

### PR DESCRIPTION
**Pull request by saitho**
Resolves #6377.

This pull request aims to implement the feature in #6377 .

I'm new to Golang. So far the feature works, however there is still some work to do:

- ~~Resolve build issues (linting etc)~~
- ~~Reimplement webhooks that are executed when target branch is changed~~
- ~~Read and adhere to Contribution guidelines~~
- ~~Squash commits into one commit that follows the existing commit style of the project~~

### CODE REVIEW

File: options/locale/locale_bg-BG.ini
Comment (lunny) [inline options/locale/locale_bg-BG.ini]: Only en-US translation should be committed.

**Review by lunny — CHANGES_REQUESTED** [review]

File: options/locale/locale_bg-BG.ini
Comment (saitho) [reply options/locale/locale_bg-BG.ini]: Ah right, translations are handled via CrowdIn. In order to change the target branch name on the already loaded page I'd need to assign a ID to the translation string (pulls.title_desc) like this: `<code id="branch_target">%[3]s</code>`

Apart from the newly introduced string in the English (and German) translation file that's the only change to the translations I've made. I'll remove these changes later and resubmit them via CrowdIn.

**Review by techknowlogick — DISMISSED** [review]
Please add migration for changes to structs

**Comment by saitho** [comment]
I'm done for now. :) Please have a look at it again @techknowlogick  @lunny :)

The only thing that bugs me is that pull requests are nested inside issues...

After clicking the "Edit" button on a pull request page the user is able to change the title (as before) and select a new target branch from a select box. Since the title belongs to issues and the target branch belongs to a pull request there will be 2 HTTP requests. One setting the title via the existing issue endpoint and one setting the target branch via a new pull request endpoint. Which is also why two web hooks will be fired. :(

**Comment by saitho** [comment]
What needs to be done in order to merge this? :)
(Apart from resolving the merge conflicts on the migrations files, obviously ^^)

**Comment by techknowlogick** [comment]
Im currently traveling so I can't review right now, but I'll dismiss my review so it doesn't block this PR.

**Comment by lunny** [comment]
please resolve the file conflicting.

**Comment by lunny** [comment]
I think there are still some work need to do when change the branch of pull request. It should affect many aspects.

**Comment by saitho** [comment]
> I think there are still some work need to do when change the branch of pull request. It should affect many aspects.

This PR is probably not going to make it in 1.9.0, is it? ;)
Maybe we could start thinking about aspects being affected by changing target branches.

File: templates/repo/issue/view_title.tmpl
Comment (saitho) [inline templates/repo/issue/view_title.tmpl:58]: As we also have this select of source and target branch on the regular pull request page, we should put it inside a partial and use that on the pull request form and here. By passing settings to the partial we'd be able to chose which field should be editable.

**Comment by saitho** [comment]
What needs to be done here in order to get this merged in 1.11.0? ;)

**Comment by lunny** [comment]
Please resolve the conflicts

**Comment by saitho** [comment]
Conflicts resolved and tested locally again.

Before changing the target branch we should check if the source branch and new target branch are equal or if a pull request already exists, similar to the validation when creating a PR.

**Comment by guillep2k** [comment]
> 
> 
> Conflicts resolved and tested locally again.
> 
> Before changing the target branch we should check if the source branch and new target branch are equal or if a pull request already exists, similar to the validation when creating a PR.

Can you add that check?

**Comment by saitho** [comment]
I started working on that.
Right now I have to figure out how I can get the repository to check if branches are equal. ^^
Also I'll have to check if there is a mechanism for showing error messages of AJAX calls to the user.
Right now the request aborts without telling the user anything. ;)

**Comment by guillep2k** [comment]
@saitho If the sha1 of the last commit in one branch and the other are the same, then the branches are equal.

**Comment by saitho** [comment]
> @saitho If the sha1 of the last commit in one branch and the other are the same, then the branches are equal.

Yup. My problem was I couldn't get the repository from pr object because I didn't call `GetBaseRepository()`. :)

**Comment by saitho** [comment]
> @saitho If the sha1 of the last commit in one branch and the other are the same, then the branches are equal.

Shouldn't that be "Branches are equal if the sha1 of the last commit in head branch is present in the base branch (not necessarily the last commit)"?
At least that's what's working for me. :)

**Comment by saitho** [comment]
The error messages (existing PR, equal branches) are now printed into a flash message.
Other errors that might occur will cause a server error (500).

Unfortunately the localization however doesn't work yet :(
failing_i18n

**Comment by guillep2k** [comment]
> Shouldn't that be "Branches are equal if the sha1 of the last commit in head branch is present in the base branch (not necessarily the last commit)"?
Oh, you mean that they have a common ancestor!

**Comment by lunny** [comment]
Please resolve the conflicts.

**Comment by saitho** [comment]
I'm still resolving conflicts. Currently the branch can only be changed if the title is also changed...
Will investigate that (compiled JS may contain console.log debugs ^^)

**Comment by saitho** [comment]
> Unfortunately the localization doesn't work yet :(

Fixed. "repo." was missing at the beginning.

So far everything works as expected.
Please review. @lunny @techknowlogick

File: modules/structs/hook.go
Comment (lafriks) [inline modules/structs/hook.go]: fixme

File: go.mod
Comment (lunny) [inline go.mod]: You should use go1.13 but not go1.12

File: models/issue_comment.go
Comment (lunny) [inline models/issue_comment.go]: The length default is `VARCHAR(255)`. Is this enough for git branch name?

File: models/pull.go
Comment (lunny) [inline models/pull.go]: I would like move this function to `services/pull/`.

File: modules/notification/webhook/webhook.go
Comment (lunny) [inline modules/notification/webhook/webhook.go]: In fact, issue also has a branch name. So maybe it's unnecessary to use a seperate struct for issue and pull request?

**Review by lunny — CHANGES_REQUESTED** [review]

File: modules/notification/webhook/webhook.go
Comment (saitho) [reply modules/notification/webhook/webhook.go]: Why would Issue have a branch name? I don't see one in the model.

File: models/issue_comment.go
Comment (saitho) [reply models/issue_comment.go]: `string` is used in the PullRequest model (see _HeadBranch_ and _BaseBranch_). I believe that should be fine?

File: templates/repo/issue/view_title.tmpl
Comment (saitho) [reply templates/repo/issue/view_title.tmpl:58]: Proposal scrapped because I changed the order of the select fields in order to match the wording in the text.

File: modules/notification/webhook/webhook.go
Comment (guillep2k) [reply modules/notification/webhook/webhook.go]: > 
> 
> Why would Issue have a branch name? I don't see one in the model.

@saitho It's `Issue.Ref`.

File: models/pull.go
Comment (saitho) [reply models/pull.go]: I wasn't able to move everything into the service because of the comments.
But it does look cleaner now.

File: modules/notification/webhook/webhook.go
Comment (saitho) [reply modules/notification/webhook/webhook.go]: Oh! I didn't know that you could assign an issue to a branch.
I added `Ref` to ChangesPayload and adjusted the usages.

File: models/issue_comment.go
Comment (lafriks) [inline models/issue_comment.go]: No need for this function, just use createComment directly

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/issue_comment.go
Comment (saitho) [reply models/issue_comment.go]: So that was changed with the refactoring of the notification system. All right, will fix that later today. :)

File: routers/repo/pull.go
Comment (lafriks) [inline routers/repo/pull.go:1022]: ctx Writen should be moved before this if and here should return not found probably

**Review by lafriks — APPROVED** [review]

File: routers/repo/pull.go
Comment (guillep2k) [inline routers/repo/pull.go]: why are these `StatusInternalServerError`? Shouldn't they be 400 errors or similar? (this and the ones below)

File: models/migrations/v112.go
Comment (lunny) [inline models/migrations/v112.go]: blank line.

File: modules/git/commit.go
Comment (lunny) [inline modules/git/commit.go:313]: We should return the error

File: routers/repo/issue.go
Comment (lunny) [inline routers/repo/issue.go]: Why override line 612?

File: services/pull/pull.go
Comment (lunny) [inline services/pull/pull.go]: You have to add a new `TestPatch` like we did on `NewPullRequest` to recheck the pull request's status.

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/pull.go
Comment (saitho) [inline models/pull.go]: This does not seem to work yet. :(
Patch is not updated. Any tips welcome :)

Comment (saitho) [reply models/pull.go]: Merge of #9302 would not require this change (or I had to remove it), right? Might not be an issue then.

Comment (lunny) [reply models/pull.go]: Maybe you can wait that PR.

**Comment by lunny** [comment]
Please resolve the conflicts.

File: routers/repo/issue.go
Comment (saitho) [reply routers/repo/issue.go]: I just double checked. Might have been needed when I started the PR months ago.
Doesn't seem to be needed anymore - I removed the lines.

**Comment by saitho** [comment]
I completely moved all the ChangeBranch functionality into the pull service.
A few bits were still in the model, which I didn't like because it was a public method and one could bypass all the important checks by calling that instead of the service. ;)

Also I now understood how patch updates work and verified that the state is updated correctly.
Please re-review @lunny

File: modules/notification/webhook/webhook.go
Comment (lunny) [inline modules/notification/webhook/webhook.go:564]: This could be moved to line 564.

File: services/pull/pull.go
Comment (lunny) [inline services/pull/pull.go]: I think we should update the database on a transaction. This should be merged with `checkAndUpdateStatus`.

Comment (saitho) [reply services/pull/pull.go]: Hmm, I get see that one transaction is better than two. I just didn't want to copy code that is already in a function. Maybe I could add a parameter to `checkAndUpdateStatus` which allows updating additional columns with that transaction?

Comment (saitho) [reply services/pull/pull.go]: > Hmm, I get see that one transaction is better than two. I just didn't want to copy code that is already in a function. Maybe I could add a parameter to `checkAndUpdateStatus` which allows updating additional columns with that transaction?

Didn't work as expected. I'll just merge the contents of that function with my transaction...

File: models/pull.go
Comment (lunny) [inline models/pull.go]: or Just `return baseCommit.HasPreviousCommit(headCommit.ID)`?

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
CI failed and one small nit https://github.com/go-gitea/gitea/pull/6488#discussion_r357921158

**Comment by lafriks** [comment]
@techknowlogick @guillep2k need your approval

**Comment by techknowlogick** [comment]
@lafriks I dismissed my review so I am no longer blocking. Only @guillep2k pending now.

**Review by guillep2k — APPROVED** [review]

**Comment by saitho** [comment]
Great to see this merged. :)
Since there was a change to the translation (added id="..."), the string was removed from translations at Crowdin. I updated the translations to include the id.

Maybe we should remove HTML from the translations so we don't have to redo the translations when the HTML changes. We could open an issue for that to discuss that. ;)

**Pull request merged**
